File: fuckApp/app/src/main/java/com/fuckapp/fragment/widget/AppAdapter.java
package com.fuckapp.fragment.widget;

import android.annotation.SuppressLint;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import com.fuckapp.R;
import com.fuckapp.fragment.model.AppModel;

import java.util.ArrayList;
import java.util.List;

import static com.fuckapp.utils.Constant.HIDE_APP;

/**
 * by y on 2016/10/31
 */

public class AppAdapter extends RecyclerView.Adapter<AppAdapter.ViewHolder> {

    private List<AppModel> tempApp = new ArrayList<>();
    private OnItemClickListener onItemClickListener = null;
    private List<AppModel> appModel = new ArrayList<>();
    private SparseBooleanArray booleanArray = new SparseBooleanArray();
    private int type = -1;

    public AppAdapter(int type) {
        this.type = type;
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.app_item, parent, false));
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(ViewHolder holder, @SuppressLint("RecyclerView") final int position) {
        if (type != -1) {
            switch (type) {
                case HIDE_APP:
                    holder.checkBox.setVisibility(View.GONE);
                    break;
                default:
                    holder.checkBox.setVisibility(View.VISIBLE);
                    break;
            }
        }
        holder.mAppIcon.setImageDrawable(appModel.get(position).getAppIcon());
        holder.mAppLable.setText("程序名：" + appModel.get(position).getAppLabel());
        holder.mPkgName.setText("包名：" + appModel.get(position).getPkgName());
        if (booleanArray.get(position)) {
            holder.checkBox.setChecked(true);
            holder.itemView.setBackgroundColor(ContextCompat.getColor(holder.itemView.getContext(), R.color.colorGray));
        } else {
            holder.checkBox.setChecked(false);
            holder.itemView.setBackgroundColor(ContextCompat.getColor(holder.itemView.getContext(), R.color.colorWhite));
        }
        if (onItemClickListener != null) {
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    onItemClickListener.onItemClick(position, appModel.get(position));
                }
            });
        }
    }

    public void refreshUI(List<AppModel> list) {
        if (list != null) {
            appModel.addAll(list);
            notifyDataSetChanged();
        }
    }

    public void removeAll() {
        if (appModel != null) {
            appModel.clear();
            notifyDataSetChanged();
        }
    }

    @Override
    public int getItemCount() {
        if (appModel != null) {
            return appModel.size();
        }
        return 0;
    }

    public void refreshCheckBox(int position) {
        if (booleanArray.get(position)) {
            booleanArray.put(position, false);
            tempApp.remove(appModel.get(position));
        } else {
            booleanArray.put(position, true);
            tempApp.add(appModel.get(position));
        }
        notifyDataSetChanged();
    }

    public List<AppModel> getTempList() {
        return tempApp;
    }

    public void resetCheckbox() {
        if (booleanArray != null) {
            booleanArray.clear();
        }
        clearTempList();
        notifyDataSetChanged();
    }

    public void clearTempList() {
        if (tempApp != null) {
            tempApp.clear();
        }
    }


    public class ViewHolder extends RecyclerView.ViewHolder {
        ImageView mAppIcon;
        TextView mAppLable;
        TextView mPkgName;
        CheckBox checkBox;

        public ViewHolder(View itemView) {
            super(itemView);
            mAppIcon = (ImageView) itemView.findViewById(R.id.imgApp);
            mAppLable = (TextView) itemView.findViewById(R.id.tvAppLabel);
            mPkgName = (TextView) itemView.findViewById(R.id.tvPkgName);
            checkBox = (CheckBox) itemView.findViewById(R.id.app_check);
        }
    }

    public interface OnItemClickListener {
        void onItemClick(int position, AppModel appInfo);
    }
}

File: FuckApp/app/src/main/java/com/fuckapp/fragment/widget/AppAdapter.kt
package com.fuckapp.fragment.widget

import android.annotation.SuppressLint
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.util.SparseBooleanArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.fuckapp.R
import com.fuckapp.fragment.model.AppModel
import com.fuckapp.utils.Constant
import kotlinx.android.synthetic.main.app_item.view.*
import java.util.*

/**
 * by y on 2016/10/31
 */

class AppAdapter(var type: Int) : RecyclerView.Adapter<AppAdapter.ViewHolder>() {

    private val tempApp = ArrayList<AppModel>()
    private var onItemClickListener: OnItemClickListener? = null
    private val appModel = ArrayList<AppModel>()
    private val booleanArray = SparseBooleanArray()

    val tempList: List<AppModel>
        get() = tempApp


    fun setOnItemClickListener(onItemClickListener: OnItemClickListener) {
        this.onItemClickListener = onItemClickListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.app_item, parent, false))
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: ViewHolder, @SuppressLint("RecyclerView") position: Int) {
        if (type != -1) {
            when (type) {
                Constant.HIDE_APP -> holder.itemView.app_check.visibility = View.GONE
                else -> holder.itemView.app_check.visibility = View.VISIBLE
            }
        }
        holder.itemView.imgApp.setImageDrawable(appModel[position].appIcon)
        holder.itemView.tvAppLabel.text = "程序名：" + appModel[position].appLabel
        holder.itemView.tvPkgName.text = "包名：" + appModel[position].pkgName
        if (booleanArray.get(position)) {
            holder.itemView.app_check.isChecked = true
            holder.itemView.setBackgroundColor(ContextCompat.getColor(holder.itemView.context, R.color.colorGray))
        } else {
            holder.itemView.app_check.isChecked = false
            holder.itemView.setBackgroundColor(ContextCompat.getColor(holder.itemView.context, R.color.colorWhite))
        }
        holder.itemView.setOnClickListener { onItemClickListener?.onItemClick(position, appModel[position]) }
    }

    fun refreshUI(list: List<AppModel>?) {
        if (list != null) {
            appModel.addAll(list)
            notifyDataSetChanged()
        }
    }

    fun removeAll() {
        appModel.clear()
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return appModel.size
    }

    fun refreshCheckBox(position: Int) {
        if (booleanArray.get(position)) {
            booleanArray.put(position, false)
            tempApp.remove(appModel[position])
        } else {
            booleanArray.put(position, true)
            tempApp.add(appModel[position])
        }
        notifyDataSetChanged()
    }

    fun resetCheckbox() {
        booleanArray.clear()
        clearTempList()
        notifyDataSetChanged()
    }

    fun clearTempList() {
        tempApp.clear()
    }


    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    interface OnItemClickListener {
        fun onItemClick(position: Int, appInfo: AppModel)
    }
}

